pkg/gitreceive/build.go

```diff
@@ -183,62 +138,94 @@ func build(conf *Config, s3Client *s3.S3, builderKey, gitSha string) error {
-		"create",
-		"-f",
-		finalManifestFileLocation,
-	)
-	if log.IsDebugging {
-		kCreateCmd.Stdout = os.Stdout
-	}
-	kCreateCmd.Stderr = os.Stderr
-	if err := run(kCreateCmd); err != nil {
+	log.Debug("Pod spec: %v", pod)
+
+	podsInterface := kubeClient.Pods(conf.PodNamespace)
+
+	newPod, err := podsInterface.Create(pod)
+	if err != nil {
 		return fmt.Errorf("creating builder pod (%s)", err)
 	}
 
-	// poll kubectl every 100ms to determine when the build pod is running
-	// TODO: use the k8s client and watch the event stream instead (https://github.com/deis/builder/issues/65)
-	for {
-		cmd := kGetCmd(conf.PodNamespace, buildPodName)
-		var out bytes.Buffer
-		cmd.Stdout = &out
-		// ignore errors
-		run(cmd)
-		outStr := string(out.Bytes())
-		if strings.Contains(outStr, "phase: Running") {
+	watcher, err := podsInterface.Watch(labels.Nothing(), fields.OneTermEqualSelector("name", pod.Name), v1Version)
+	if err != nil {
+		return fmt.Errorf("watching events for builder pod startup")
+	}
+
+	ch := watcher.ResultChan()
+	for evt := range ch {
+		if evt.Type == watch.Added {
+			watcher.Stop()
 			break
-		} else if strings.Contains(outStr, "phase: Failed") {
-			return fmt.Errorf("build pod %s entered phase: Failed", buildPodName)
+		} else if evt.Type == watch.Error {
+			watcher.Stop()
+			return fmt.Errorf("builder pod failed to launch with ERROR")
 		}
-		time.Sleep(100 * time.Millisecond)
 	}
 
-	// get logs from the builder pod
-	kLogsCmd := exec.Command(
-		"kubectl",
-		fmt.Sprintf("--namespace=%s", conf.PodNamespace),
-		"logs",
-		"-f",
-		buildPodName,
-	)
-	kLogsCmd.Stdout = os.Stdout
-	if err := run(kLogsCmd); err != nil {
-		return fmt.Errorf("running %s to get builder logs (%s)", strings.Join(kLogsCmd.Args, " "), err)
+	req := kubeClient.Get().Namespace(conf.PodNamespace).Name(newPod.Name).Resource("pods").SubResource("log").VersionedParams(
+		&api.PodLogOptions{
+			Follow:   true,
+			Previous: true,
+		}, api.Scheme)
+
+	rc, err := req.Stream()
+	if err != nil {
+		return fmt.Errorf("attempting to stream logs (%s)", err)
+	}
+	defer rc.Close()
+
+	_, err = io.Copy(os.Stdout, rc)
+	if err != nil {
+		return fmt.Errorf("fetching builder logs (%s)", err)
 	}
 
 	// poll the s3 server to ensure the slug exists
-	// TODO: time out looking
+	timeout := time.After(5 * time.Minute)
+	tick := time.Tick(100 * time.Millisecond)
 	for {
-		exists, err := storage.ObjectExists(s3Client, bucketName, pushObjKey)
-		if err != nil {
-			return fmt.Errorf("Checking if object %s/%s exists (%s)", bucketName, pushObjKey, err)
-		}
-		if exists {
-			break
+		select {
+		case <-timeout:
+			return fmt.Errorf("Timed out waiting for object in storage. Aborting build...")
+		case <-tick:
+			exists, err := storage.ObjectExists(s3Client, bucketName, pushObjKey)
+			if err != nil {
+				return fmt.Errorf("Checking if object %s/%s exists (%s)", bucketName, pushObjKey, err)
+			}
+			if exists {
+				break
+			}
```

Comment:
I like this, but it seems out-of-scope for this patch. Can you move it into a different pull request?